fvm/environment/unsafe_random_generator.go

```diff
@@ -76,14 +79,24 @@ func (gen *unsafeRandomGenerator) seed() {
 		// header ID. The random number generator will be used by the
 		// UnsafeRandom function.
 		id := gen.blockHeader.ID()
-		source := rand.NewSource(int64(binary.BigEndian.Uint64(id[:])))
-		gen.rng = rand.New(source)
+		// extract the entropy from `id` and expand it into the required seed
+		hkdf := hkdf.New(func() hash.Hash { return sha256.New() }, id[:], nil, nil)
+		seed := make([]byte, random.Chacha20SeedLen)
```

Comment:
Also, could you split the changes out into multiple PRs.  It make reviewing much easier.

Reply:
> the seed/key generated by hkdf is not deterministic.

Any reference for this statement?

Btw since determinicity of unsafeRandom is the desired behaviour, there should be a test to make sure updates don't mess up with it. From my first look, there isn't any. I'll add one

Reply:
try re-running the example in the hkdf golang package.  it clearly generates different keys each time

Reply:
> could you split the changes out into multiple PRs.

This makes sense. The reason it's all in the same PR is that it was initially meant to update go version to 1.20 (https://github.com/onflow/flow-go/pull/4050), and most of the changes were needed in the same PR to make CI pass. I'll try to split it.

Reply:
> try re-running the example in the hkdf golang package. it clearly generates different keys each time

That's because the example is setting the salt (input to hkdf) to a random value at each run. You can disable the `rand.Read` part and you'd get the same outputs. Salts are used to diversify the outputs of hkdf even when used with the same seed.
In general, KDFs (hkdf is an instance) are deterministic.

Reply:
Updated the tests here (new determinicity test and improved randomness check) [77d83f0](https://github.com/onflow/flow-go/pull/4052/commits/77d83f0d1087d49fe56a3c6d3bc914527b3b89a5)

Reply:
discussion continued in https://github.com/onflow/flow-go/pull/4067